File: 5week/근우/3.step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
from collections import deque
from inspect import trace


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        startNode,destNode = None,None
        root.parent = None
        def dfs(root):
            nonlocal startNode,destNode
            if not root:
                return
            
            if root.val == startValue:
                startNode = root
            if root.val ==destValue:
                destNode = root
            
            if root.left:
                root.left.parent = root
                dfs(root.left)
            if root.right:
                root.right.parent = root
                dfs(root.right)
        dfs(root)
        is_visit = [False for _ in range(10**5+1)]
        que = deque([(startNode,'')]) # curNode, tracing_routes
        while que:
            node, traces = que.popleft()
            if node == destNode:
                return traces
            for next_node, dir in [(node.parent,'U'), (node.left,'L'), (node.right,'R')]:
                if next_node and not is_visit[next_node.val]:
                    is_visit[next_node.val] = True
                    que.append((next_node,traces+dir))

File: 5week/남규/3. Step-By-Step Directions From a Binary Tree Node to Another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traverse(TreeNode* curr, string& s, int v1, int v2, string &start, string &end){
        if(curr->left){
            s.push_back('L');
            traverse(curr->left, s, v1, v2, start, end);
            s.pop_back();
        }
        
        if(curr->right){
            s.push_back('R');
            traverse(curr->right, s, v1, v2, start, end);
            s.pop_back();
        }
        
        if(v1 == curr->val){
            start = s;
            return;
        }
        
        if(v2 == curr->val){
            end = s;
            return;
        }
    }
    
    string getDirections(TreeNode* root, int startValue, int destValue) {
        string s = "";
        string e = "";
        string tmp = "";
        traverse(root, tmp, startValue, destValue, s, e);
        int lca = 0;
        for(; lca < min(s.size(), e.size()); lca++){
            if(s[lca] != e[lca]){
                break;
            }
        }
        return string(s.size() - lca, 'U')  + e.substr(lca);
    }
};
